preop.cpp

```diff
@@ -405,43 +406,21 @@ char *handle_preop_define (const char *ptr) {
  * return NULL if there's no matching path
  */
 
-char *full_path (const char *filename) {
-	list_t *dir;
-	char *full_path;
-#ifdef WIN32
-	if (is_abs_path(filename) && (GetFileAttributesW(widen(filename).c_str()) != INVALID_FILE_ATTRIBUTES))
-#else
-	if (is_abs_path(filename) && (access (filename, R_OK) == 0))
-#endif
-		return strdup (filename);
-	
-	dir = include_dirs;
-	full_path = NULL;
-	do if (dir) {
-		expand_buf_t *eb = eb_init (-1);
-		
-		eb_append (eb, (char *) dir->data, -1);
-		eb_append (eb, "/", 1);
-		eb_append (eb, filename, -1);
-		free (full_path);
-		full_path = eb_extract (eb);
-		fix_filename (full_path);
-		eb_free (eb);
-		dir = dir->next;
-#ifdef WIN32
-	} while (GetFileAttributes(widen(full_path).c_str()) == INVALID_FILE_ATTRIBUTES && dir);
-#else
-	} while (access (full_path, R_OK) && dir);
-#endif
+char *full_path (const char *filename_bytes) {
+	std::error_code fs_err;
+	const std::filesystem::path filename(filename_bytes);

```

Comment:
This doesn't do UTF-8 conversion, unlike the code you're replacing. Since you're building C++17, you should use std::filesystem::u8path() to generate the path object.

Reply:
The diff on this commit is a little misleading, since I added the UTF-8 handling in d56556d3e1d11c283592518b6a6a52799eab004f but don't think that was actually correct. The `basic_string` path constructor seems correct because we haven't actually checked what the encoding of our bytestrings is.

I think this means it'll end up using the ANSI encoding, but I'm not sure there's a better solution and if the process is running with the UTF-8 codepage then we get exactly what we want.

Reply:
Ah, that makes sense. I think the ideal "correct" solution is accepting wide string CLI arguments and converting them to UTF-8, but that brings back some crustiness. Using UTF-8 for filenames specified in source files does seem like the most consistent thing to do, however...

Reply:
Maybe the best thing to do is embed an application manifest in Windows builds that sets UTF-8 mode (`activeCodePage` = `UTF-8`), then we can ignore the rest of the Window oddities.

Reply:
Added an application manifest in 0e961dac927077d45ec67c4b00a57414785d33c2 that requests UTF-8 mode that I think solves this, at least for sufficiently recent versions of Windows.